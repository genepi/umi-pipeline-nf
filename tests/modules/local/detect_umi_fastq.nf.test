nextflow_process {

    name "Test Process DETECT_UMI_FASTQ"
    script "modules/local/umi_processing/detect_umi_fastq.nf"
    process "DETECT_UMI_FASTQ"

    test("Should run without failures") {

        when {
            params{
                use_context = false
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode02", "raw", "$baseDir/tests/input/detect_umi/barcode02/lpa2645_filtered.fastq"],
                    ["barcode03", "raw", "$baseDir/tests/input/detect_umi/barcode03/lpa2645_filtered.fastq"],
                    ["barcode04", "raw", "$baseDir/tests/input/detect_umi/barcode04/lpa2645_filtered.fastq"],
                    ["barcode05", "raw", "$baseDir/tests/input/detect_umi/barcode05/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }
    test("Should detect the same UMIs - short adapter") {

        when {
            params{
                adapter_length = 50
                use_context = false
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode02", "lpa2645", "$baseDir/tests/input/detect_umi/barcode02/lpa2645_filtered.fastq"],
                    ["barcode03", "lpa2645", "$baseDir/tests/input/detect_umi/barcode03/lpa2645_filtered.fastq"],
                    ["barcode04", "lpa2645", "$baseDir/tests/input/detect_umi/barcode04/lpa2645_filtered.fastq"],
                    ["barcode05", "lpa2645", "$baseDir/tests/input/detect_umi/barcode05/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
            with(process.out.umi_extract_fastq){
                assert path(get(0).get(2)).md5 == "d41d8cd98f00b204e9800998ecf8427e"
                assert path(get(1).get(2)).md5 == "ef06dacd82cc76a3e1797fafed348d97"
                assert path(get(2).get(2)).md5 == "75c1c38a134cf1270f5e94d3ea938006"
                assert path(get(3).get(2)).md5 == "fe23b244bdfdfe5f62e2488c6acff007"
            }
        }

    }

    test("Should detect the same UMIs - long adapter") {

        when {
            params{
                adapter_length = 500
                use_context = false
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode02", "lpa2645", "$baseDir/tests/input/detect_umi/barcode02/lpa2645_filtered.fastq"],
                    ["barcode03", "lpa2645", "$baseDir/tests/input/detect_umi/barcode03/lpa2645_filtered.fastq"],
                    ["barcode04", "lpa2645", "$baseDir/tests/input/detect_umi/barcode04/lpa2645_filtered.fastq"],
                    ["barcode05", "lpa2645", "$baseDir/tests/input/detect_umi/barcode05/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
            with(process.out.umi_extract_fastq){
                assert path(get(0).get(2)).md5 == "d41d8cd98f00b204e9800998ecf8427e"
                assert path(get(1).get(2)).md5 == "ef06dacd82cc76a3e1797fafed348d97"
                assert path(get(2).get(2)).md5 == "75c1c38a134cf1270f5e94d3ea938006"
                assert path(get(3).get(2)).md5 == "fe23b244bdfdfe5f62e2488c6acff007"
            }
        }

    }

    test("Should detect UMIs") {

        when {
            params{
                adapter_length = 250
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode_context", "lpa2645", "$baseDir/tests/input/detect_umi/barcode_context/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }
    }

    test("Should detect the UMIs using context primer") {

        when {
            params{
                adapter_length = 250
                use_context = true
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode_context", "lpa2645", "$baseDir/tests/input/detect_umi/barcode_context/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
            with(process.out.umi_extract_fastq){
                assert path(get(0).get(2)).md5 == "d98f53ef8fb52f5ca0cffda16e10c5b5"
            }
        }

    }

    test("Should output empty file") {

        when {
            params {
                use_context = false
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode02", "lpa2645", "$baseDir/tests/input/detect_umi/barcode02/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
            with(process.out.umi_extract_fastq){
                assert path(get(0).get(2)).readLines().size() == 0
            }
        }

    }

    test("Should not find any UMI") {

        when {
            params{
                adapter_length = 20
                use_context = false
            }
            process {
                """
                input[0] = Channel.of(
                    ["barcode02", "lpa2645", "$baseDir/tests/input/detect_umi/barcode02/lpa2645_filtered.fastq"],
                    ["barcode03", "lpa2645", "$baseDir/tests/input/detect_umi/barcode03/lpa2645_filtered.fastq"],
                    ["barcode04", "lpa2645", "$baseDir/tests/input/detect_umi/barcode04/lpa2645_filtered.fastq"],
                    ["barcode05", "lpa2645", "$baseDir/tests/input/detect_umi/barcode05/lpa2645_filtered.fastq"],
                )
                input[1] = file( "$baseDir/.nf-test/tests/")
                input[2] = "raw"
                input[3] = file("$baseDir/bin/extract_umis.py")
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
            with(process.out.umi_extract_fastq){
                assert path(get(0).get(2)).readLines().size() == 0
                assert path(get(1).get(2)).readLines().size() == 0
                assert path(get(2).get(2)).readLines().size() == 0
                assert path(get(3).get(2)).readLines().size() == 0
            }
        }

    }

}
