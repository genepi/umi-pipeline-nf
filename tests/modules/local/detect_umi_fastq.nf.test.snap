{
    "Should detect the same UMIs - long adapter": {
        "content": [
            {
                "0": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.fastq:md5,ef06dacd82cc76a3e1797fafed348d97"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.fastq:md5,75c1c38a134cf1270f5e94d3ea938006"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.fastq:md5,fe23b244bdfdfe5f62e2488c6acff007"
                    ]
                ],
                "1": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.tsv:md5,f1fb453e4c7e2496b9ef2bd3940a3579"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.tsv:md5,669511d3a919f063813db044a8a42d65"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.tsv:md5,353a7a314c1a8f46bb05b74cc035c64d"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.fastq:md5,ef06dacd82cc76a3e1797fafed348d97"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.fastq:md5,75c1c38a134cf1270f5e94d3ea938006"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.fastq:md5,fe23b244bdfdfe5f62e2488c6acff007"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.tsv:md5,f1fb453e4c7e2496b9ef2bd3940a3579"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.tsv:md5,669511d3a919f063813db044a8a42d65"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.tsv:md5,353a7a314c1a8f46bb05b74cc035c64d"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T16:03:58.490129719"
    },
    "Should detect the UMIs using context primer": {
        "content": [
            {
                "0": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.fastq:md5,d98f53ef8fb52f5ca0cffda16e10c5b5"
                    ]
                ],
                "1": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.tsv:md5,242202b6b70f9216e3594432722bfbc8"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.fastq:md5,d98f53ef8fb52f5ca0cffda16e10c5b5"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.tsv:md5,242202b6b70f9216e3594432722bfbc8"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T15:35:24.97172069"
    },
    "Should detect the same UMIs - short adapter": {
        "content": [
            {
                "0": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.fastq:md5,ef06dacd82cc76a3e1797fafed348d97"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.fastq:md5,75c1c38a134cf1270f5e94d3ea938006"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.fastq:md5,fe23b244bdfdfe5f62e2488c6acff007"
                    ]
                ],
                "1": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.tsv:md5,f1fb453e4c7e2496b9ef2bd3940a3579"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.tsv:md5,669511d3a919f063813db044a8a42d65"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.tsv:md5,353a7a314c1a8f46bb05b74cc035c64d"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.fastq:md5,ef06dacd82cc76a3e1797fafed348d97"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.fastq:md5,75c1c38a134cf1270f5e94d3ea938006"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.fastq:md5,fe23b244bdfdfe5f62e2488c6acff007"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.tsv:md5,f1fb453e4c7e2496b9ef2bd3940a3579"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.tsv:md5,669511d3a919f063813db044a8a42d65"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.tsv:md5,353a7a314c1a8f46bb05b74cc035c64d"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T15:54:25.636651399"
    },
    "Should detect UMIs": {
        "content": [
            {
                "0": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.fastq:md5,f02454a788bf0ce6bbff1a05b64afa75"
                    ]
                ],
                "1": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.tsv:md5,c1a3bfcf4bcf7894f12e9b3038c62c66"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.fastq:md5,f02454a788bf0ce6bbff1a05b64afa75"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode_context",
                        "lpa2645",
                        "extracted_umis_barcode_context_1.tsv:md5,c1a3bfcf4bcf7894f12e9b3038c62c66"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T16:04:07.498446952"
    },
    "Should output empty file": {
        "content": [
            {
                "0": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T15:35:32.631398014"
    },
    "Should not find any UMI": {
        "content": [
            {
                "0": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.tsv:md5,1d7cb5b915aa09b2dc37531f66222a45"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.tsv:md5,fb3a528714e818653113163369c6dcf8"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.tsv:md5,0dba9d0ed906743cc8150b8bea054b35"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode02",
                        "lpa2645",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "lpa2645",
                        "extracted_umis_barcode03_2.tsv:md5,1d7cb5b915aa09b2dc37531f66222a45"
                    ],
                    [
                        "barcode04",
                        "lpa2645",
                        "extracted_umis_barcode04_3.tsv:md5,fb3a528714e818653113163369c6dcf8"
                    ],
                    [
                        "barcode05",
                        "lpa2645",
                        "extracted_umis_barcode05_4.tsv:md5,0dba9d0ed906743cc8150b8bea054b35"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T15:35:44.486305974"
    },
    "Should run without failures": {
        "content": [
            {
                "0": [
                    [
                        "barcode02",
                        "raw",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "raw",
                        "extracted_umis_barcode03_2.fastq:md5,ef06dacd82cc76a3e1797fafed348d97"
                    ],
                    [
                        "barcode04",
                        "raw",
                        "extracted_umis_barcode04_3.fastq:md5,75c1c38a134cf1270f5e94d3ea938006"
                    ],
                    [
                        "barcode05",
                        "raw",
                        "extracted_umis_barcode05_4.fastq:md5,fe23b244bdfdfe5f62e2488c6acff007"
                    ]
                ],
                "1": [
                    [
                        "barcode02",
                        "raw",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "raw",
                        "extracted_umis_barcode03_2.tsv:md5,f1fb453e4c7e2496b9ef2bd3940a3579"
                    ],
                    [
                        "barcode04",
                        "raw",
                        "extracted_umis_barcode04_3.tsv:md5,669511d3a919f063813db044a8a42d65"
                    ],
                    [
                        "barcode05",
                        "raw",
                        "extracted_umis_barcode05_4.tsv:md5,353a7a314c1a8f46bb05b74cc035c64d"
                    ]
                ],
                "umi_extract_fastq": [
                    [
                        "barcode02",
                        "raw",
                        "extracted_umis_barcode02_1.fastq:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ],
                    [
                        "barcode03",
                        "raw",
                        "extracted_umis_barcode03_2.fastq:md5,ef06dacd82cc76a3e1797fafed348d97"
                    ],
                    [
                        "barcode04",
                        "raw",
                        "extracted_umis_barcode04_3.fastq:md5,75c1c38a134cf1270f5e94d3ea938006"
                    ],
                    [
                        "barcode05",
                        "raw",
                        "extracted_umis_barcode05_4.fastq:md5,fe23b244bdfdfe5f62e2488c6acff007"
                    ]
                ],
                "umi_extract_fastq_stats": [
                    [
                        "barcode02",
                        "raw",
                        "extracted_umis_barcode02_1.tsv:md5,ad5e1bbeb6edb594841e6cfd3172d20c"
                    ],
                    [
                        "barcode03",
                        "raw",
                        "extracted_umis_barcode03_2.tsv:md5,f1fb453e4c7e2496b9ef2bd3940a3579"
                    ],
                    [
                        "barcode04",
                        "raw",
                        "extracted_umis_barcode04_3.tsv:md5,669511d3a919f063813db044a8a42d65"
                    ],
                    [
                        "barcode05",
                        "raw",
                        "extracted_umis_barcode05_4.tsv:md5,353a7a314c1a8f46bb05b74cc035c64d"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-10-07T16:03:32.90071395"
    }
}